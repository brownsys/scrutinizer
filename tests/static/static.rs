#![allow(dead_code)]

mod mutable_static {
    static mut GLOBAL_VEC: Vec<u32> = vec![];

    #[doc = "impure"]
    fn leak_into_static(a: u32) {
        unsafe {
            GLOBAL_VEC.push(a);
        }
    }
}

mod mutation_from_static {
    struct PureIncrementer;

    impl PureIncrementer {
        fn inc(&self, a: usize) -> usize {
            a + 1
        }
    }

    struct ImpureIncrementer;

    impl ImpureIncrementer {
        fn inc(&self, a: usize) -> usize {
            println!("{}", a);
            a + 1
        }
    }

    static PURE_INCREMENTER: PureIncrementer = PureIncrementer {};
    static IMPURE_INCREMENTER: ImpureIncrementer = ImpureIncrementer {};

    #[doc = "pure"]
    fn pure_call_from_static(a: usize) -> usize {
        PURE_INCREMENTER.inc(a)
    }

    #[doc = "impure"]
    fn impure_call_from_static(a: usize) -> usize {
        IMPURE_INCREMENTER.inc(a)
    }
}

fn main() {}
