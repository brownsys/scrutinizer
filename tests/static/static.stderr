error: static analysis was not able to verify the purity of the region
  --> static.rs:7:5
   |
7  | /     fn leak_into_static(a: u32) {
8  | |         unsafe {
9  | |             GLOBAL_VEC.push(a);
10 | |         }
11 | |     }
   | |_____^
   |
   = help: consider using sandbox or privacy region
   = note: `#[deny(scrutinizer_lint)]` on by default

error: static analysis was not able to verify the purity of the region
  --> static.rs:41:5
   |
41 | /     fn impure_call_from_static(a: usize) -> usize {
42 | |         IMPURE_INCREMENTER.inc(a)
43 | |     }
   | |_____^
   |
   = help: consider using sandbox or privacy region
